django/middleware/csrf.py

```diff
@@ -196,6 +227,25 @@ def _set_token(self, request, response):
             # Set the Vary header since content varies with the CSRF cookie.
             patch_vary_headers(response, ('Cookie',))
 
+    def _origin_verified(self, request):
+        request_origin = request.META['HTTP_ORIGIN']
+        good_origin = '%s://%s' % (
+            'https' if request.is_secure() else 'http',
+            request.get_host(),
+        )
+        if request_origin == good_origin:
+            return True
+        if request_origin in self.allowed_origins_exact:
+            return True
+        parsed_origin = urlparse(request_origin)
+        request_scheme = parsed_origin.scheme
+        request_netloc = parsed_origin.netloc
+        for allowed_scheme, allowed_host in self.allowed_origin_subdomains:
+            if (request_scheme == allowed_scheme and
+                    is_same_domain(request_netloc, allowed_host)):
+                return True
+        return False
```

Comment:
By the way, something else that just occurred to me. It seems like this `for` loop could be simplified or eliminated if `self.allowed_origin_subdomains` were to instead cache a dict mapping scheme to hosts. For example, something like:

```python
return any(
    is_same_domain(request_netloc, host) for host in
        self.allowed_origin_subdomains.get(request_scheme, ())
)
```

Reply:
done